Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/PaymentMethodTokenizationViewModel.swift

```diff
@@ -309,7 +309,227 @@ class PaymentMethodTokenizationViewModel: NSObject, PaymentMethodTokenizationVie
+                
+        let settings: PrimerSettingsProtocol = DependencyContainer.resolve()
+        
+        if Primer.shared.flow.internalSessionFlow.vaulted {
+            Primer.shared.delegate?.tokenAddedToVault?(paymentMethod)
+            self.handleSuccess()
+        } else if settings.isManualPaymentHandlingEnabled {
+            
+            PrimerDelegateProxy.onTokenizeSuccess(paymentMethod, resumeHandler: self)
+            PrimerDelegateProxy.onTokenizeSuccess(paymentMethod, { err in
+                if let err = err {
+                    self.handleFailedTokenizationFlow(error: err)
+                } else {
+                    self.handleSuccessfulTokenizationFlow()
+                }
+            })
+
+        } else {
+                        
+            guard let paymentMethodTokenString = paymentMethod.token else {
+                
+                DispatchQueue.main.async {
+                    let paymentMethodTokenError = PrimerError.invalidValue(key: "resumePaymentId", value: "Payment method token not valid", userInfo: ["file": #file, "class": "\(Self.self)", "function": #function, "line": "\(#line)"])
+                    ErrorHandler.handle(error: paymentMethodTokenError)
+                    self.handleErrorBasedOnSDKSettings(paymentMethodTokenError)
+                }
+                
+                return
+            }
+            
+            firstly {
+                self.handleCreatePaymentEvent(paymentMethodTokenString)
+            }
+            .done { paymentResponse -> Void in
+                
+                guard let paymentResponse = paymentResponse else {
+                    return
+                }
+
+                self.resumePaymentId = paymentResponse.id
+                
+                if paymentResponse.status == .pending, let requiredAction = paymentResponse.requiredAction {
+                    self.handle(newClientToken: requiredAction.clientToken)
+                } else {
+                    let checkoutData = CheckoutData(payment: CheckoutDataPayment(from: paymentResponse))
+                    Primer.shared.delegate?.primerDidCompleteCheckoutWithData?(checkoutData)
+                    self.handleSuccess()
+                }
+            }
+            .catch { error in
+                self.handleErrorBasedOnSDKSettings(error)
+            }
+        }
+    }
+        
+    // Raise Primer will create Payment event
+    
+    internal func handlePrimerWillCreatePaymentEvent(_ paymentMethodData: PaymentMethodData) -> Promise<Void> {
+        return Promise { seal in
+            let checkoutPaymentMethodType = CheckoutPaymentMethodType(type: paymentMethodData.type.rawValue)
+            let checkoutPaymentMethodData = CheckoutPaymentMethodData(type: checkoutPaymentMethodType)
+            Primer.shared.delegate?.primerWillCreatePaymentWithData?(checkoutPaymentMethodData, decisionHandler: { paymentCreationDecision in
```

Comment:
What happens if the merchant hasn't implemented the `primerWillCreatePaymentWithData`? Won't it hang?

Reply:
Addressed the completion within PrimerDelegateProxy ✅